### ISSUE TITLE
DistributedDataParallel trainer interferes with GANTrainer.switch()

### ISSUE THREAD

**Issue by philtrade** [fastai/fastai#2496]
**Describe the bug**

Training wgan in Distributed Data Parallel mode would crash, because it cannot find the `switch()` method in the `DistributedDataParallel` object.

**Provide your installation details**
```
fastai v1, python 3.7, PyTorch v1.2 and v1.4, 3 cuda GPUs (gtx 1080 ti)
platform      : Linux-4.15.0-62-generic-x86_64-with-Ubuntu-18.04-bionic
```

**To Reproduce**
Run the scripted version of `lesson7-wgan` notebook in DDP mode, even with only 1 GPU:  https://gist.github.com/philtrade/cb9953079a99afc522fa777574f2549d
by `python -m torch.distributed.launch --nproc_per_node=1 wgan_ddp.py`
will produce:
```text
  File "/home/ndim1/Dropbox/fastai/fastai/basic_train.py", line 98, in fit
    cb_handler.on_epoch_begin()
  File "/home/ndim1/Dropbox/fastai/fastai/callback.py", line 273, in on_epoch_begin
    self('epoch_begin')
  File "/home/ndim1/Dropbox/fastai/fastai/callback.py", line 251, in __call__
    for cb in self.callbacks: self._call_and_update(cb, cb_name, **kwargs)
  File "/home/ndim1/Dropbox/fastai/fastai/callback.py", line 241, in _call_and_update
    new = ifnone(getattr(cb, f'on_{cb_name}')(**self.state_dict, **kwargs), dict())
  File "/home/ndim1/Dropbox/fastai/fastai/vision/gan.py", line 134, in on_epoch_begin
    self.switch(self.gen_mode)
  File "/home/ndim1/Dropbox/fastai/fastai/vision/gan.py", line 154, in switch
    self.model.switch(gen_mode)
  File "/usr/local/lib/python3.7/dist-packages/torch/nn/modules/module.py", line 576, in __getattr__
    type(self).__name__, name))
AttributeError: 'DistributedDataParallel' object has no attribute 'switch'
```

**Expected behavior**
Training should proceed, e.g.:
```
epoch     train_loss  valid_loss  gen_loss  disc_loss  time    
Epoch 1/3 : |███---------------------------------------------------------------| 5.92% [10/169 00:26<07:00 -0.1393
```

**Triage**
The primary problem is fairly straight-forward.  `GANTrainer.switch()` attempts to call the `switch()` method on the `GANModule`, its `self.model`, in this line: 
 https://github.com/fastai/fastai/blob/4e630af8b416e8ea1553f12f3d0dcd4a7fd7ea86/fastai/vision/gan.py#L153
`self.model` points to `self.learn.model` when `GANTrainer` was constructed.

But in DDP mode upon `DistributedTrainer.on_train_begin()`,`self.learn.model` becomes  PyTorch's `DistributedDataParallel` object, which hides the original `GANModule` inside its `.module` member, therefore the above call cannot reach the intended `GANModule` object.

**Proposed Fix**
Similar to how `GANTrainer` remembers the `generator` and `critic` in the constructor:
https://github.com/fastai/fastai/blob/4e630af8b416e8ea1553f12f3d0dcd4a7fd7ea86/fastai/vision/gan.py#L78
Remember the original `switch` method as well:
```
        self.generator,self.critic = self.model.generator,self.model.critic
+       self.switch_model = self.model.switch
```
then change https://github.com/fastai/fastai/blob/4e630af8b416e8ea1553f12f3d0dcd4a7fd7ea86/fastai/vision/gan.py#L153
to `self.switch_model(gen_mode)`

I have tested this in PyTorch v1.4, and I'll be happy to draft a PR for this.

--
Note 1: PyTorch v1.2 has a bug [(issue 19840)](https://github.com/pytorch/pytorch/issues/19840) that further requires `DistributedDataParallel()` be called with `...,find_unused_parameters=True` in `fastai.distributed.DistributedTrainer.on_train_begin()`.   [PyTorch v1.4 fixed it](https://github.com/pytorch/pytorch/pull/19901) though.

Note 2: With the above machinery fixed, training `wgan` in DDP has other problems.  In short, the losses (train, gen, disc) don't seem to drop from epoch to epoch, and training doesn't speed up despite use of multiple GPUs.  I'm investigating, and will raise as separate issues if warranted.

**Comment by sgugger** [comment]
YEs, a PR with this would be most welcome. Thanks!

**Comment by sgugger** [comment]
Closed by #2498
